src/cache/event.rs

```diff
@@ -38,113 +38,52 @@ use crate::model::user::{CurrentUser, OnlineStatus};
-
-                        channel.recipient.id
-                    };
-
-                    if let Some(u) = cache.users.get(&user_id) {
-                        channel.recipient = u.clone();
-                    }
-
-                    channel.id
-                };
+        let old_channel = cache
+            .guilds
+            .get_mut(&self.channel.guild_id)
+            .and_then(|mut g| g.channels.insert(self.channel.id, self.channel.clone()));
 
-                cache.private_channels.insert(id, channel.clone()).map(Channel::Private)
-            },
-        }
+        cache.channels.insert(self.channel.id, self.channel.guild_id);
+        old_channel
     }
 }
 
 impl CacheUpdate for ChannelDeleteEvent {
     type Output = Vec<Message>;
 
     fn update(&mut self, cache: &Cache) -> Option<Vec<Message>> {
-        match &self.channel {
-            Channel::Guild(channel) => {
-                let (guild_id, channel_id) = (channel.guild_id, channel.id);
+        let (channel_id, guild_id) = (self.channel.id, self.channel.guild_id);
 
-                cache.channels.remove(&channel_id);
-
-                cache.guilds.get_mut(&guild_id).map(|mut g| g.channels.remove(&channel_id));
-            },
-            Channel::Private(channel) => {
-                let id = { channel.id };
-
-                cache.private_channels.remove(&id);
-            },
-        };
+        cache.channels.remove(&channel_id);
+        cache.guilds.get_mut(&guild_id).map(|mut g| g.channels.remove(&channel_id));
 
         // Remove the cached messages for the channel.
-        cache
-            .messages
-            .remove(&self.channel.id())
-            .map(|(_, messages)| messages.into_values().collect())
+        cache.messages.remove(&channel_id).map(|(_, messages)| messages.into_values().collect())
     }
 }
 
 impl CacheUpdate for ChannelUpdateEvent {
-    type Output = Channel;
-
-    fn update(&mut self, cache: &Cache) -> Option<Channel> {
-        let old_channel = cache.channel(self.channel.id());
+    type Output = GuildChannel;
 
-        match &self.channel {
-            Channel::Guild(channel) => {
-                cache.channels.insert(channel.id, channel.guild_id);
+    fn update(&mut self, cache: &Cache) -> Option<GuildChannel> {
+        cache.channels.insert(self.channel.id, self.channel.guild_id);
 
-                cache
-                    .guilds
-                    .get_mut(&channel.guild_id)
-                    .map(|mut g| g.channels.insert(channel.id, channel.clone()));
-            },
-            Channel::Private(channel) => {
-                if let Some(mut c) = cache.private_channels.get_mut(&channel.id) {
-                    c.clone_from(channel);
-                }
-            },
-        }
-
-        old_channel
+        cache
+            .guilds
+            .get_mut(&self.channel.guild_id)
+            .and_then(|mut g| g.channels.insert(self.channel.id, self.channel.clone()))
     }
 }
```

Comment:
Having trouble reviewing this because the diff is quite messy. But, I'm assuming changing `.map` to `.and_then` works because `HashMap::insert` returns the old value wrapped in `Some` (and `None` if there is no old value)?

It's clever, but maybe it's worth adding a comment to that effect so that other people can better understand what's happening without much effort.

Reply:
I kind of agree with you here, but looking at the whole module sees this trick used quite a couple of times, so I don't think it's necessary to comment here.